### ISSUE TITLE
Web Support

### ISSUE THREAD

**Issue by connectety** [renefloor/flutter_cache_manager#122]
Please add support for Flutter web.
Maybe this can be achieved using localStorage.

**Comment by renefloor** [comment]
The problem is not storing the files, but that it uses sqflite to store the cache info and that doesn't support Flutter Web yet: https://github.com/tekartik/sqflite/issues/212

I'll look into this whether a switch to Moor or Hive https://github.com/renefloor/flutter_cache_manager/issues/121 is a good move. We'll have to migrate all the data, so I really have to study all the pros and cons before making a decision.

**Comment by connectety** [comment]
I looked into it a little bit more.
if I _understand everything correctly_ and did not _forget any type of storage_,
localStorage has a maximum of 10 MB per origin and
cookies have a maximum of 4 KB per domain and
indexDB can get quite big but is not persistent.

In my opinion this is not enough space to cache files.
But webbrowser cache files themselves quite good.

As a hotfix I would propose, not caching files for web, but giving a warning and / or exception, which has to be dismissed.

**Comment by renefloor** [comment]
@connectety I'm not completely sure what you propose. Do you mean that if the platform is Web the library just uses NetworkImage instead of CachedImage? Devs will expect it to work differently.

**Comment by connectety** [comment]
> use NetworkImage instead of CachedImage

Yes, for example.

>  Devs will expect it to work differently.

I know, and this is not the best idea I have ever had.
I do not stand behind it completely, but I don't see any way to save content and a database, because the storage provided by web browsers is just too small.

Hopefully, I have forgotten some kind of storage.

And also I want to use this as a Hotfix, which can only be used after dismissing an exception or setting a flag or something. So that people will understand that it does not work as expected.
This idea should still somewhat work because there is already pretty good caching in the web and browsers.

Even if this objectively bad solution gets dismissed, I wanted to point out the Problem.

**Comment by alextekartik** [comment]
I started a fork (https://github.com/alextekartik/flutter_cache_manager and https://github.com/alextekartik/flutter_cached_network_image) where I experimented caching for the web by abstracting the file system (using fs_shim) and the storage (using idb_shim). It is definitely not perfect and a better solution could be found but I think having a way to abstract the database and file system into flutter_cache_manager could allow anyone to use their preferred storage (hive, moor, sqlite, indexed db....)

But with CORS and web browser already doing some caching, the web is a different experience that could require a specific and different network caching mechanism

**Comment by IchordeDionysos** [comment]
Why not use service workers for caching? 🤔 

This is what the web does really good out of the box. (In newer browser versions :D)
Also with the Cache API, you can save the most files with at least 50MB storage.

This might need some refactoring of this package to use sqflite for mobile and service workers for web (maybe using the plugin_platform_interface)

Useful links:
https://developer.mozilla.org/en-US/docs/Web/API/Cache

**Comment by deakjahn** [comment]
Until the support arrives, could it be made to fail silently? :-)

    await DefaultCacheManager().emptyCache();

simply never returns for me in a web app. Yes, it's easy to wrap it into an `if (!kIsWeb)`, just saying...

**Comment by renefloor** [comment]
It doesn't cache on disk at the moment, but as @deakjahn asked it at least doesn't fail. It nicely returns the file which is just downloaded from the internet.

**Comment by deakjahn** [comment]
Actually, I store my files with the Cache API (not necessarily these one but generally). Do you need some code so that you don't need to look it up yourself? Fairly straightforward.

**Comment by renefloor** [comment]
Yes it is always good to see some code, but it looks really straight forward indeed.

The main challenge I think at the moment is adapting the cache manager to separate the current storage from the web api.

**Comment by deakjahn** [comment]
I have an app-embedded platform plugin that simply offers a file storage interface for my app and gets translated to either plain old `dart.io` or Cache API, depending on the platform. It's tailored to my specific needs, so it's not perfect for immediate re-use without some tweaking. Basically, whenewer the normal file IO speaks about folders, I use separate cache files on the web and treat a path like "foo/bar" as an item stored with the name "bar" inside a cache file named "foo". Nothing more nested than this, this was enough for my purposes.

So, the interface itself is this:

```dart
Future<bool> fileSave(String path, Uint8List contents, {String mime, bool flush = false});
Future<bool> fileSaveString(String path, String contents, {String mime, bool flush = false});
Future<Uint8List> fileRead(String path, {String mime});
Future<String> fileReadString(String path, {String mime});
Future<bool> fileExists(String path);
Future<int> fileSize(String path);
Future<bool> fileDelete(String path);
Future<bool> fileCopy(String src, String dest);
Future<bool> folderCreate(String path, {bool recursive = false});
Future<bool> folderDelete(String path, {bool recursive = false});
Future<bool> folderEmpty(String path);
Future<bool> folderExists(String path);
Future<List<String>> folderList(String path);
```

Not that you would need all of them, especially not the ones that store a string (basically, a text file).

**Comment by deakjahn** [comment]
The implementation is relatively simple:

```dart
import 'dart:html' as html;
import 'dart:js';
import 'dart:js_util' as util;

const String _BINARY = 'application/octet-stream';
const String _TEXT = 'text/plain';

Future<bool> fileSave<T>(String path, T contents, String mime) async {
  final parts = path.splitFirst('/');
  return util.promiseToFuture<bool>(_nativeCacheSave(parts.first, parts.last, contents, mime));
}

Future<T> fileRead<T>(String path, String mime) async {
  final parts = path.splitFirst('/');
  if (mime == _BINARY) {
    final buffer = await util.promiseToFuture<ByteBuffer>(await _nativeCacheRead(parts.first, parts.last));
    return buffer.asUint8List() as T;
  } else if (mime == _TEXT)
    return await util.promiseToFuture<String>(_nativeCacheReadString(parts.first, parts.last)) as T;
  else
    throw UnsupportedError(mime);
}

Future<bool> fileExists(String path) async {
  final parts = path.splitFirst('/');
  return util.promiseToFuture<bool>(_nativeCacheExists(parts.first, parts.last));
}

Future<int> fileSize(String path) {
  final parts = path.splitFirst('/');
  return util.promiseToFuture<int>(_nativeCacheSize(parts.first, parts.last));
}

Future<bool> fileDelete(String path) async {
  final parts = path.splitFirst('/');
  return util.promiseToFuture<bool>(_nativeCacheDelete(parts.first, parts.last));
}

Future<bool> fileCopy(String src, String dest) async {
  final parts1 = src.splitFirst('/');
  final parts2 = dest.splitFirst('/');
  return util.promiseToFuture<bool>(_nativeCacheCopy(parts1.first, parts1.last, parts2.first, parts2.last));
}

Future<bool> folderCreate(String path) async {
  final parts = path.splitFirst('/');
  await html.window.caches.open(parts.first);
  return true;
}

Future<bool> folderDelete(String path) async {
  final parts = path.splitFirst('/');
  return await html.window.caches.delete(parts.first);
}

Future<bool> folderEmpty(String path) async {
  await folderDelete(path);
  await folderCreate(path);
  return true;
}

Future<bool> folderExists(String path) async {
  final parts = path.splitFirst('/');
  return await html.window.caches.has(parts.first);
}

Future<List<String>> folderList(String path) async {
  final files = await util.promiseToFuture<List<dynamic>>(_nativeCacheList(path));
  return files.map((f) => f.toString()).toList();
}
```

with a simple helper:

```dart
extension Splits on String {
  List<String> splitFirst(Pattern pattern) {
    final parts = split(pattern);
    return [
      if (parts.isNotEmpty) parts.first,
      if (parts.length > 1) parts.sublist(1).join(pattern),
    ];
  }
}
```

**Comment by deakjahn** [comment]
The native (JavaScript, that is) implementations are:

```dart
@JS('cacheSave')
external bool _nativeCacheSave(String name, String path, dynamic contents, String mime);

@JS('cacheRead')
external JsObject _nativeCacheRead(String name, String path);

@JS('cacheReadString')
external String _nativeCacheReadString(String name, String path);

@JS('cacheCopy')
external bool _nativeCacheCopy(String name1, String src, String name2, String dest);

@JS('cacheExists')
external bool _nativeCacheExists(String name, String path);

@JS('cacheSize')
external int _nativeCacheSize(String name, String path);

@JS('cacheDelete')
external bool _nativeCacheDelete(String name, String path);

@JS('cacheList')
external JsArray _nativeCacheList(String name);
```

```js
cacheSave: async function(name, path, contents, mime) {
  var cache = await window.caches.open(name);
  var blob = new Blob([contents], {type : mime});
  var options = {headers: {'Content-Type': mime, 'Content-Length': contents.length}};
  await cache.put(path, new Response(blob, options));
  return true;
},

cacheRead: async function(name, path) {
  var cache = await window.caches.open(name);
  var options = {ignoreSearch: true, ignoreMethod: true, ignoreVary: true};
  var data = await cache.match(path, options);
  return data.arrayBuffer();
},

cacheReadString: async function(name, path) {
  var cache = await window.caches.open(name);
  var options = {ignoreSearch: true, ignoreMethod: true, ignoreVary: true};
  var data = await cache.match(path, options);
  return data.text();
},

cacheCopy: async function(name1, src, name2, dest) {
  var cache1 = await window.caches.open(name1);
  var options = {ignoreSearch: true, ignoreMethod: true, ignoreVary: true};
  var data = await cache1.match(src, options);
  var cache2 = await window.caches.open(name2);
  await cache2.put(dest, data);
  return true;
},

cacheSize: async function(name, path) {
  var cache = await window.caches.open(name);
  var options = {ignoreSearch: true, ignoreMethod: true, ignoreVary: true};
  var data = await cache.match(path, options);
  return parseInt(data.headers.get('Content-Length'));
},

cacheExists: async function(name, path) {
  var cache = await window.caches.open(name);
  return await cache.match(path) != undefined;
},

cacheDelete: async function(name, path) {
  var cache = await window.caches.open(name);
  return await cache.delete(path);
},

cacheList: async function(name) {
  var cache = await window.caches.open(name);
  var files = await cache.keys();
  return files.map(f => new URL(f.url).pathname);
},
```

But this supposes that you already have a JS file that you can put them into. If you don't, I can also show you how I normally attach one. :-)

**Comment by deakjahn** [comment]
Oh yes, and of course, you need to make sure you have the support:

```js
cacheAvailable = window.caches != undefined;
```

You do with any sane browser you might encounter today but still... :-)

**Comment by renefloor** [comment]
> But this supposes that you already have a JS file that you can put them into.

If you use dart:html in a Flutter app, isn't this automatically added?

```
  var blob = new Blob([contents], {type : mime});
  var options = {headers: {'Content-Type': mime, 'Content-Length': contents.length}};
  await cache.put(path, new Response(blob, options));
```
You basically take the content from the response and put them in a new response. Why wouldn't you just directly add the response from the server?

> you need to make sure you have the support

Yes of course, thanks for that as well. Especially for older iOS devices (iOS 10 and lower) this is important.

**Comment by deakjahn** [comment]
No, nothing is automatically added, you need the `import`, anyway.

The Cache API only stores responses, not plain files. If you want to store _arbitrary_ data, you have to create a `Response` whose contents is the data you need to store. But yes, in your case, you might be able to simply store the original response, yes, Note that, as I said in the beginning, I use this as a transparent cross platform storage solution, not necessarily to store actual HTTP responses.

**Comment by deakjahn** [comment]
Yes, come to think of it, you can make it quite a bit simpler than my code. The Cache API is originally conceived to store responses keyed by the actual URL. I had to bend over backwards a little bit to make this fit a more universal file storing need.

**Comment by deakjahn** [comment]
Sure but I can't yet see where to connect. As I peek into `flutter_cache_manager`, I can see no specific web plugin code. Do you plan to put it somewhere here? Are some parts already written or nothing yet?

**Comment by deakjahn** [comment]
@renefloor Just to mention. There are some significant differences between 2.3.0-beta.1 and 2.3.0-rc. In release mode, the beta works flawlessly for me but the rc does not. I couldn't yet look into what exactly it is (I checked and the sources are radically different, of course, between the two).

**Comment by renefloor** [comment]
@deakjahn Are you using SKIA or normal web platform?

I'm separating the web and io file systems here: 
https://github.com/Baseflow/flutter_cache_manager/pull/193/files#diff-052689d714d3d840d1f79fe281252833
https://github.com/Baseflow/flutter_cache_manager/pull/193/files#diff-7b8a439543b2745e7c428f80e25e27d7

The implementation is chosen by defining which config implementation to use here: https://github.com/Baseflow/flutter_cache_manager/pull/193/files#diff-a0ac6e0ea1a2b0d4c6776f859429b621

**Comment by deakjahn** [comment]
@renefloor Both. The app normally uses Skia but when I just start it in debug mode and want to check something that's not really related to graphics rendering, that's normally the old web (only because we can't yet automatically start debug in Skia mode, actually; if we could, I probably never used DomCanvas at all).

But does the caching funciton depend on that, really?

**Comment by renefloor** [comment]
@deakjahn The caching mechanism doesn't really depend on that, but the image rendering does. 
The default HTML ImageElement doesn't work on Skia, so that is why (as far as I'm aware) NetworkImage doesn't work at the moment in Skia. See also: https://github.com/flutter/flutter/issues/54010
When you use Skia you should just download the image separately and render it yourself. However, when doing an HttpGet you have to manage CORS headers, so you bump into different problems. 

In 2.3.0-beta I used HTML ImageElement, in 2.3.0-beta.1 I used an HTTP Get by making this package compatible. In 2.3.0-rc I added both with a flag which option to use. When you set imageRenderMethodForWeb to ImageRenderMethodForWeb. HttpGet (not the default) it should work the same as 2.3.0-beta.1.

**Comment by deakjahn** [comment]
That's confimed. In DomCanvas, both settings work. In CanvasKit, HtmlImage fails and HttpGet works. Then I stay on the `rc` package.

So, where should things go forward now? Is HttpGet meant to be the way to go? Do we need to bother with the other one? Is it important to check while HtmlImage fails (I supposed you already done that, anyway, with the Flutter bug mentioned above). Or leave all that as it is and go on to caching?

**Comment by deakjahn** [comment]
Oh, I can see now why I couldn't see. The web support you mentioned is not yet in the package, it's only in a PR? Then I'll need to learn how to make that happen over here, too. :-) There is an Apply Patch in AS, so probably that's what it is.

**Comment by renefloor** [comment]
>The web support you mentioned is not yet in the package, it's only in a PR? 
Yes, my goal was to remove all dependencies on dart:io in the web version, but that is actually really hard to do, so I'm not sure how far I will go. At least the config object is a good idea I think.

>So, where should things go forward now? Is HttpGet meant to be the way to go? Do we need to bother with the other one? 
For now I will support both. When using an HTML ImageElement the image always loads (no need to bother with CORS or other shit) and the caching is handled by the browser. I think this is the right way for most users, that is why it is the default.

For more power users that want more control on the caching mechanism using the HttpGet is a way with more options. In making a choice between them I follow the general discussion in the Flutter repo. 
Topics of interest: 
Use of Skia with NetworkImage: https://github.com/flutter/flutter/issues/54010
Http headers for NetworkImage: https://github.com/flutter/flutter/issues/57187
(Old) discussion about CORS: https://github.com/flutter/flutter/issues/45955

>Or could it be somehow made work if I forked the package on GitHub?
You can fork from the branch and make a PR to merge it into that branch again. I guess I can also give you write access to that specific branch.

**Comment by deakjahn** [comment]
OK, I forked that one. Are there any specific internal dependencies I have to follow? The very first attempt resulted in:

    Target dart2js failed: Exception: ../../flutter-beta/.pub-cache/hosted/pub.dartlang.org/cached_network_image-2.3.0-rc/lib/src/image_provider/_image_provider_web.dart:119:34:
    Error: The method 'readAsBytes' isn't defined for the class 'CacheFile'.

So I probably need to adapt related modules as well?

**Comment by renefloor** [comment]
@deakjahn Yes, the PR is not final yet. You can also wait till I finish more of this project.

**Comment by deakjahn** [comment]
But do you have a local copy that displays and you could zip up and send? I assume that the task at hand would be to rewrite `cache_file_web.dart`, to get rid of the `file` dependency and to read and write the Cache API here, am I correct? It's not a problem, I guess then, if other parts are still being rewritten.

**Comment by renefloor** [comment]
@deakjahn I didn't have any working version that was compatible with CachedNetworkImage.
This commit fixes the build and makes sure CachedNetworkImage runs: https://github.com/Baseflow/flutter_cache_manager/pull/193/commits/8cc0dcc7c65c4f35f276555130b2d90efc7357fb

There are indeed 2 files that should be changed.
`file_system_web` should store and retrieve files from the CacheStore. 
`cache_file_web` should not use `'package:file/file.dart'` for files, but `'dart:html'`. However, for that we need to make a custom implementation for methods as `exists()` or `delete()` as those files are only 'in memory' files if I'm correct.

We also need to changed the IOSink. I guess we can make a method that accepts a stream.

We also need to replace the sqflite CacheObjectProvider, but I'll do that myself.

**Comment by deakjahn** [comment]
No, they're not in memory, they are stored on the disk.

It seems that I'm up and running with that latest source, so I start to took into the caching now.

**Comment by deakjahn** [comment]
So, in spite of what we mentioned earlier, we don't store raw queries and responses but separate files. So, that Blob approach will remain. :-)

**Comment by deakjahn** [comment]
@renefloor Now I have a more fundamental problem. The setup you currently have doesn't really lend itself to cache storage. The files you normally store are independent, so you only have the creating of the folder at the FileSystem level and all reading, writing, deleting is done with the individual files. However, think about cache as being on the folder level, not the level of the individual files. You have to ask the folder (the cache) whether an entry exists, not the file itself. You have to ask the folder (the cache) to delete the entry, not the file itself. If I place the cache entry itself into CacheFile, I can actually read the contents but nothing else: no write, no delete or similar.

**Comment by deakjahn** [comment]
@renefloor But I just can't decide yet what you intend with createParent() and openWrite()...

**Comment by deakjahn** [comment]
@renefloor So, I have my latest version up there. Right now the problems seems to be that writeAsBytes() and readAsBytes() are called in parallel. We try to read before we actually had a chance to write completely. This would probably require your help because I can't yet clearly see how you actually call these, it's a bit complicated at first sight. :-)

Apart from that, the images get stored all right. When you run an example, open the Dev Console of Chrome, go to the Application tab, open the Cache Storage in the menu list, you'll see the libCachedImageData entry there and the stored images inside. Don't expect a preview because the MIME is a generic image/*. I'm not sure it would be worth the trouble to store the exact one.

**Comment by renefloor** [comment]
Great work @deakjahn. I'll check it as soon as I have time for it :)

**Comment by deakjahn** [comment]
Right now `readAsBytes()` gets called and returns the stored image but it doesn't actually display. I don't yet know where it's called from because it's not called directly, some other async operations reaches it, probably via a stream or similar. If I can spot where the call comes from, I'll put a few `print()`'s there to see what happens...

**Comment by renefloor** [comment]
@deakjahn sorry for the late response, had a very busy time.
I switched to your branch, but I can't get the example working (when building for chrome).
At first I had a lot of errors like this:
![image](https://user-images.githubusercontent.com/15101411/87670234-b4647080-c76f-11ea-9931-c49780f0a526.png)
Then I just added casts like `promise2 as JsObject`. The example was missing CacheObject.path, so I removed that.

Now I just get "Failed to compile application."

Shouldn't there also be a JS file somewhere?

**Comment by deakjahn** [comment]
No JS file. I know I mentioned the JS file solution above but for these few and simple calls, I finally thought this would be simpler.

**Comment by renefloor** [comment]
@deakjahn I found `readAsBytes()` is not a valid function in the file package. However, are those casts correct or should they just work as in your branch?

**Comment by deakjahn** [comment]
To tell you frankly, I kinda lost track, I try to pick it up now. I replaced the current 1.4.1 with this version now. I run it with `ImageRenderMethodForWeb.HttpGet` now, no compilation error (Flutter beta 1.19.0-4.3.pre). The simple test program runs but doesn't display the image. The image gets stored in the cache, though.

**Comment by deakjahn** [comment]
I seem to remember that I kept the `file` package dependency because of `IOSink` but nothing else is actually used from that.

**Comment by renefloor** [comment]
@deakjahn You can easily replace IOSink with `StreamSink<List<int>>`, which is basically what it is.

**Comment by deakjahn** [comment]
If I put print()s into both ˙writeAsBytes()` and `readAsBytes()` of the web `CacheFile`, both get called. So, the underlying mechanism seems to work for me, just that it doesn't yet displays for some reason.

I'll try the `IOSink` modification.

**Comment by deakjahn** [comment]
This is my very simple testbed:

```dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Cache Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Cache Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  final String title;

  MyHomePage({Key key, this.title}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Image(
          image: CachedNetworkImageProvider(
            'https://picsum.photos/200/300',
            imageRenderMethodForWeb: ImageRenderMethodForWeb.HttpGet,
          ),
        ),
      ),
    );
  }
}
```

**Comment by renefloor** [comment]
Which version of the cache manager are you using now? The latest version of your branch doesn't compile for me, but do you use that?

**Comment by deakjahn** [comment]
The simple `StreamSink` replacement will not float but sink :-) because `CacheFile` won't implement it. I'll leave is as it is for now.

I have the source stashed away here that, supposedly, is the same I uploaded to GitHub. Should be your special branch with my modifications in two or three files.

**Comment by deakjahn** [comment]
The files around here: https://github.com/deakjahn/flutter_cache_manager/tree/feature/separate-html-from-io/flutter_cache_manager/lib/src/storage/file_system

**Comment by deakjahn** [comment]
Because I didn't want to install it on my computer separately (I have running projects that use your current one), I didn't push-pull it with Git. Just uploaded the three files manually, so all my commits say "Add files via upload". In fact, only three files were changed at all: cache_file_web.dart, file_system_web.dart and _config_web.dart. All the rest remained untouched.

**Comment by deakjahn** [comment]
Actually, all I have here is the testbed I quoted above, depending on cached_network_image-2.3.0-rc and this modified flutter_cache_manager. And it just compiles for me. :-)

Ok. Could you also tell me which line the compiler says this about?

**Comment by deakjahn** [comment]
Is this the `return await _getBlobData(blob);`?

**Comment by deakjahn** [comment]
What about leaving out the `await` there? It's not even required at all:

      final promise2 = data.callMethod('blob', []);
      final blob = await _toFuture<html.Blob>(promise2);
      return _getBlobData(blob);

**Comment by renefloor** [comment]
![image](https://user-images.githubusercontent.com/15101411/87686155-76be1280-c784-11ea-937e-16291645cde8.png)
It's all these red lines.

But if your local version is with the changes, why don't push-pull it with Git?

**Comment by renefloor** [comment]
Btw, I did commit the IOSink change, because I think we should not depend on it if we don't have to: https://github.com/Baseflow/flutter_cache_manager/pull/193/commits/095e9ba965fd94949066e9f3a3c35a9046222242

**Comment by deakjahn** [comment]
Oh, the _toFuture's. I don't know if that's got anything to do with it but you probably use VS Code, don't you? I have Android Studio. It shouldn't matter, I just got used to it because I used it earlier, anyway, to do native Android stuff and it was just comfortable to keep using it.

**Comment by deakjahn** [comment]
Yes, but the problem with this IOSink modification is that I used it as: 

    class CacheFile implements def.CacheFile, IOSink

It wasn't very elegant, I know, but under the circumstances... :-) `openWrite()` should return something that can be used later.

**Comment by deakjahn** [comment]
https://api.flutter.dev/flutter/dart-io/File/openWrite.html

What shall we return here if we do away with IOSink completely?

**Comment by deakjahn** [comment]
One problem at a time. :-))

What does this version do for you?

    Future<T> _toFuture<T>(dynamic promise) {
      final completer = Completer<T>();
      (promise as js.JsObject).callMethod('then', [
        (result) => completer.complete(result)
      ]);
      return completer.future;
    }

It still works over here.

**Comment by renefloor** [comment]
![image](https://user-images.githubusercontent.com/15101411/87688042-c00f6180-c786-11ea-91eb-62783313b7c8.png)

I use Intellij, which is the same as Android Studio, but that shouldn't matter for the analysis.

**Comment by deakjahn** [comment]
Yes, it shouldn't. Now I started to see something. I didn't open the whole package with AS or whatever IDE before because I didn't see the need. I just modified those three files in the flutter pub cache directory and compiled my test app (the test app **was** under the IDE, of course, just not your package).

If I **open** your package in the IDE, I get the same errors. However, if I just compile an app that **uses** your package, there are no errors at all, it compiles and runs. Go figure.

**Comment by deakjahn** [comment]
Anyway, the modified `_toFuture()` above got rid of half of the errors for me, so it might be a good start.

**Comment by deakjahn** [comment]
I start to get the feeling that this is `pedantic` stuff. :-)

**Comment by deakjahn** [comment]
Here we go: https://github.com/deakjahn/flutter_cache_manager/commit/66d74d2bc4bf8ebe0aff454cae616178f3ad6fff

**Comment by deakjahn** [comment]
However, still with `IOSink` because that should be solved first. You can simply remove it from the IO but I don't know how you want to handle it on the Web.

**Comment by wer-mathurin** [comment]
Any status on this?

**Comment by renefloor** [comment]
Not at the moment.

**Comment by mrtdeh** [comment]
any update on this?

**Comment by davemg3** [comment]
Hello,
Is this still scheduled to add web support on flutter cache manager?
thks

**Comment by tomekit** [comment]
I was wondering if there are still plans to add a Web support. It would greatly improve: `flutter_cached_network_image` caching capabilities.

**Comment by duck-dev-go** [comment]
Any update on this?

**Comment by kszczek** [comment]
This feature can now be easily implemented using pure Dart with no supporting JS code thanks to the recently released `package:web`. Here's a quick demo:

```dart
import 'dart:js_interop';

import 'package:flutter/material.dart';
import 'package:web/web.dart' show Response, window;

void main() {
  runApp(const CacheApiDemo());
}

class CacheApiDemo extends StatelessWidget {
  const CacheApiDemo({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Cache API Demo',
      home: Scaffold(
        appBar: AppBar(title: Text('Cache API Demo'), centerTitle: true),
        body: Center(
          child: SizedBox(
            width: 300,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.stretch,
              spacing: 8.0,
              children: <Widget>[
                FilledButton(onPressed: _read, child: Text('Read')),
                FilledButton(onPressed: _write, child: Text('Write')),
                FilledButton(onPressed: _delete, child: Text('Delete')),
              ],
            ),
          ),
        ),
      ),
    );
  }

  Future<void> _read() async {
    final cache = await window.caches.open('testCache').toDart;
    final result = await cache.match('testKey'.toJS).toDart;

    if (result != null) {
      print(await result.text().toDart);
    } else {
      print('testKey missing!');
    }
  }

  Future<void> _write() async {
    final cache = await window.caches.open('testCache').toDart;
    await cache.put('testKey'.toJS, Response('testValue'.toJS)).toDart;
  }

  Future<void> _delete() async {
    final cache = await window.caches.open('testCache').toDart;
    await cache.delete('testKey'.toJS).toDart;
  }
}
```

The Cache API allows us to store an [enourmous](https://developer.mozilla.org/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria#other_web_technologies) amount of data (up to 60% of total disk space in Chrome and Safari!).

There is a small issue to think about though:

```dart
cache.put('testKey'.toJS, Response('testValue'.toJS)).toDart;
cache.put('testKey'.toJS, Response('anotherValue'.toJS)).toDart;
```

I haven't read the spec so I'm not sure if the `.put()` method guarantees that the call order will reflect the order in which the values will be persisted, but in case it does not guarantee that, then I suggest we use the synchronous local storage API as a mutex, this way we'll also be protected from simultaneous cross-tab writes.

**Comment by nicolarigoni** [comment]
@ben-milanko hi, any update on this? Thanks

**Comment by ben-milanko** [comment]
Hey @nicolarigoni ,

I've made a PR for this change https://github.com/Baseflow/flutter_cache_manager/pull/504

But looking at the repo the last update was 2024, so I'm not sure if this is still actively being maintained.

If you would like to use it, you can include the below in your pubspec yaml.

```yaml
  cached_network_image:
    git:
      url: https://github.com/Railway-Engineering-Solutions/flutter_cached_network_image.git
      ref: develop
  flutter_cache_manager: #^3.1.2
    git:
      url: https://github.com/Railway-Engineering-Solutions/flutter_cache_manager.git
      ref: develop
      path: flutter_cache_manager
```
